File: LLD/Design-Patterns/Flyweight/code/Flyweight.cpp
// Flyweight Pattern — sharing tree type data (intrinsic) across many tree instances.
// Compile: g++ -std=c++14 Flyweight.cpp -o flyweight
// Run:     ./flyweight

#include <iostream>
#include <string>
#include <unordered_map>
#include <memory>

// The Flyweight: shared, immutable intrinsic state.
class TreeType {
public:
    TreeType(std::string name, std::string color, std::string texture)
        : name_(std::move(name)), color_(std::move(color)), texture_(std::move(texture)) {}

    void render(int x, int y) const {
        // x, y are extrinsic state, passed in rather than stored here.
        std::cout << "Rendering " << name_ << " (" << color_ << ", " << texture_ << ") at (" << x << "," << y << ")" << std::endl;
    }

private:
    std::string name_, color_, texture_;
};

// The Flyweight Factory: caches one TreeType per unique combination.
class TreeTypeFactory {
public:
    static TreeType* getTreeType(const std::string& name, const std::string& color, const std::string& texture) {
        std::string key = name + "-" + color + "-" + texture;
        auto it = cache_.find(key);
        if (it == cache_.end()) {
            std::cout << "Creating new TreeType for: " << key << std::endl;
            auto result = cache_.emplace(key, std::make_unique<TreeType>(name, color, texture));
            return result.first->second.get();
        }
        return it->second.get();
    }

    static size_t cacheSize() { return cache_.size(); }

private:
    static std::unordered_map<std::string, std::unique_ptr<TreeType>> cache_;
};

std::unordered_map<std::string, std::unique_ptr<TreeType>> TreeTypeFactory::cache_;

// Tree only stores extrinsic state + a pointer to its shared TreeType.
class Tree {
public:
    Tree(int x, int y, TreeType* type) : x_(x), y_(y), type_(type) {}

    void render() const { type_->render(x_, y_); }

private:
    int x_, y_;
    TreeType* type_;
};

int main() {
    Tree forest[] = {
        Tree(10, 20, TreeTypeFactory::getTreeType("Oak", "Green", "Rough")),
        Tree(30, 40, TreeTypeFactory::getTreeType("Oak", "Green", "Rough")),
        Tree(50, 60, TreeTypeFactory::getTreeType("Pine", "DarkGreen", "Smooth")),
    };

    for (const auto& tree : forest) tree.render();

    std::cout << "Trees created: 3, TreeType flyweights cached: " << TreeTypeFactory::cacheSize() << std::endl;

    return 0;
}

File: v1/LLD/Design-Patterns/Flyweight/code/Flyweight.java
// Flyweight Pattern — sharing tree type data (intrinsic) across many tree instances.
// Compile: javac Flyweight.java
// Run:     java Flyweight

import java.util.HashMap;
import java.util.Map;

// The Flyweight: shared, immutable intrinsic state.
class TreeType {
    private final String name;
    private final String color;
    private final String texture;

    TreeType(String name, String color, String texture) {
        this.name = name;
        this.color = color;
        this.texture = texture;
    }

    void render(int x, int y) {
        // x, y are extrinsic state, passed in rather than stored here.
        System.out.println("Rendering " + name + " (" + color + ", " + texture + ") at (" + x + "," + y + ")");
    }
}

// The Flyweight Factory: caches one TreeType per unique combination.
class TreeTypeFactory {
    private static Map<String, TreeType> cache = new HashMap<>();

    static TreeType getTreeType(String name, String color, String texture) {
        String key = name + "-" + color + "-" + texture;
        if (!cache.containsKey(key)) {
            System.out.println("Creating new TreeType for: " + key);
            cache.put(key, new TreeType(name, color, texture));
        }
        return cache.get(key);
    }

    static int cacheSize() {
        return cache.size();
    }
}

// Tree only stores extrinsic state + a reference to its shared TreeType.
class Tree {
    private int x, y;
    private TreeType type;

    Tree(int x, int y, TreeType type) {
        this.x = x;
        this.y = y;
        this.type = type;
    }

    void render() {
        type.render(x, y);
    }
}

public class Flyweight {
    public static void main(String[] args) {
        Tree[] forest = new Tree[]{
            new Tree(10, 20, TreeTypeFactory.getTreeType("Oak", "Green", "Rough")),
            new Tree(30, 40, TreeTypeFactory.getTreeType("Oak", "Green", "Rough")),
            new Tree(50, 60, TreeTypeFactory.getTreeType("Pine", "DarkGreen", "Smooth")),
        };

        for (Tree tree : forest) {
            tree.render();
        }

        System.out.println("Trees created: " + forest.length + ", TreeType flyweights cached: " + TreeTypeFactory.cacheSize());
    }
}
